pkg/store/proxy.go

```diff
@@ -323,7 +323,52 @@ func storeMatches(s Client, mint, maxt int64, matchers ...storepb.LabelMatcher)
 func (s *ProxyStore) LabelNames(ctx context.Context, r *storepb.LabelNamesRequest) (
 	*storepb.LabelNamesResponse, error,
 ) {
-	return nil, status.Error(codes.Unimplemented, "not implemented")
+	var (
+		warnings []string
+		names    [][]string
+		mtx      sync.Mutex
+		g, gctx  = errgroup.WithContext(ctx)
+	)
+
+	stores, err := s.stores(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Unknown, err.Error())
+	}
+	for _, st := range stores {
+		st := st
+		g.Go(func() error {
+			resp, err := st.LabelNames(gctx, &storepb.LabelNamesRequest{
+				PartialResponseDisabled: r.PartialResponseDisabled,
+			})
+			if err != nil {
+				err = errors.Wrapf(err, "fetch label names from store %s", st)
+				if r.PartialResponseDisabled {
+					return err
+				}
+
+				mtx.Lock()
+				warnings = append(warnings, errors.Wrap(err, "fetch label names").Error())
```

Comment:
```suggestion
				warnings = append(warnings, errors.Sprintf("fetch label names: %s", err))
```

Reply:
I think maybe using `Wrap` would be more make sense than `Sprintf`.

Reply:
IMO not this creates a wrapped erro object and then unwraps it using `Error()` which is confusing.

Reply:
I think `warnings = append(warnings, err.Error())` is actually enough ;p

Reply:
Done